hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+import           Prelude                  hiding (tanh)
+import           System.Environment
+import           System.IO.Unsafe
+import qualified Torch.Autograd           as A
+import qualified Torch.Device             as D
+import qualified Torch.DType              as D
+import qualified Torch.Functions          as D
+import qualified Torch.NN                 as A
+import qualified Torch.Tensor             as D
+import qualified Torch.TensorFactories    as D
+import           Torch.Typed
+import           Torch.Typed.Factories
+import           Torch.Typed.Native       (expand, lstm)
+import           Torch.Typed.NN
+import           Torch.Typed.Tensor
+
+
+-- | A specification for a long, short-term memory layer.
+--
+data LSTMSpec device (dir :: Directionality)
+        (dtype :: D.DType)
+        (numLayers:: Nat)
+        (inputDim :: Nat)
+        (hiddenDim :: Nat) =
+    LSTMSpecZerosInit DropoutSpec -- ^ Weights drawn from Xavier-Uniform with zeros-value
+                                  --   initialized biases and cell states.
+    | LSTMSpec                        -- ^ Weights drawn from Xavier-Uniform
+                                      --   with zeros-value initialized biases
+                                      --   and user-provided cell states.
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial values of the memory cell
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial values of the hidden state
+        DropoutSpec
+
+    | LSTMSpecLearnedInit             -- ^ Weights drawn from Xavier-Uniform
+                                      --   with zeros-value initialized biases
+                                      --   and learned cell states.
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial (learnable)
+                                      -- values of the memory cell
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial (learnable)
+                                      -- values of the hidden state
+        DropoutSpec
+    deriving (Show, Generic)
+
+
+-- | LSTMParams
+-- Input-to-hidden, hidden-to-hidden, and bias parameters for a mulilayered
+-- (and optionally) bidirectional LSTM.
+--
+data LSTMParams (device :: (D.DeviceType, Nat)) (dtype :: D.DType) (numDirections :: Nat) (inputDim :: Nat) (hiddenSize :: Nat) (numLayers :: Nat) where
+    LSTMLayer1 ::Parameter device dtype '[4 * hiddenSize, inputDim]
+                    -> (Parameter device dtype '[4 * hiddenSize, hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> LSTMParams device dtype numDirections inputDim hiddenSize 1
+    LSTMLayerK ::(1 <= numLayers)
+                    => LSTMParams device dtype numDirections inputDim hiddenSize numLayers
```

Comment:
instead of nesting layers like this, we could also use an `HList`. what do you think?

Reply:
I can see there's a good argument for the `HList` but I'm inclined to leave it for another time (like when `Generics` become more valuable).  

In general, I find the inductive GADTs to be simpler in developing the invariants but once you know what they are, I guess going to type families and constraints isn't that bad.  

(Though I have to say, figuring out the `Overlapping` pragmas and the needed constraints on that singleton helper class was a hard-won struggle I would have preferred to avoid . . . )